### ISSUE TITLE
Align handling of hyphenated words with LIAAD/yake

### ISSUE THREAD

**Issue by bunny-therapist** [quesurifn/yake-rust#20]
This comes up in the google-text test, since it contains the word "competition-centric", but this is most easily demonstrated using a short example text:

"I am a competition-centric person! I really like competitions. Every competition is a hoot!"

Extracting 10 keywords with yake-rust (my branch, but this problem is not handled in any branch) returns (this is a set, not ordered - does not matter for this example):
```
{'person', 'competition', 'hoot', 'centric'}
```
whereas LIAAD/yake returns
```
{'person', 'competition', 'hoot', 'competition-centric'}
```
In other words, yake-rust treats "competition" and "centric" as two words, whereas LIAAD/yake keeps "competition-centric" as one word.

I tried removing `-` from the punctuation symbols, but that does not fix it. I tried to figure out if it had anything to do with the unicode segmentation, but that investigation is still ongoing.

**Comment by bunny-therapist** [comment]
Based on my playing with `split_word_bounds` in the rust playground, it does not seem like unicode-segmentation splits "competition-centric".

**Comment by xamgore** [comment]
Tokenization approaches differ, yeah. I bet `segtok` handles English texts better than `unicode-segmentation`. Not much of a choice, though. https://github.com/huggingface/tokenizers is a relatively novel crate, haven't tried it yet.

As for `Yake`, it should accept the tokenized input as argument.

**Comment by bunny-therapist** [comment]
So you would have to tokenize everything yourself? Wouldn't that make the crate a lot less useful? Or are you suggesting that as an option?

**Comment by xamgore** [comment]
Tokenize yourself or use the predefined fallback from `yake`.

It makes sense, as language specific tokenizers show better performance. Check out [Razdel](https://github.com/natasha/razdel?tab=readme-ov-file#tokens) for `ru` or [sentok](https://github.com/fnl/syntok?tab=readme-ov-file#sentence-segmentation-and-word-tokenization) for `en`, `de`.

**Comment by bunny-therapist** [comment]
Ok, as long as there is still a default fallback, I agree.

I am all for configurability, I just want to make sure it is an easy drop-in replacement for LIAAD/yake in most cases.

**Comment by xamgore** [comment]
Could you please share the list of tokens which python impl produces for google text? `[[word, ..], ..]`

**Comment by bunny-therapist** [comment]
Assuming that the list of tokens requested is the thing generated in [this line](https://github.com/LIAAD/yake/blob/master/yake/datarepresentation.py#L50):
```
 self.sentences_str = [ [w for w in split_contractions(web_tokenizer(s)) if not (w.startswith("'") and len(w) > 1) and len(w) > 0] for s in list(split_multi(text)) if len(s.strip()) > 0]     
```
Then you can find them in this attached file:
[sentences_str.json](https://github.com/user-attachments/files/17586728/sentences_str.json)

**Comment by bunny-therapist** [comment]
From there it creates a list of "unique terms". This involves e.g. the plural-normalization. You can find those here: 
[terms.json](https://github.com/user-attachments/files/17586750/terms.json)

**Comment by bunny-therapist** [comment]
This is the word tokenizer used in LIAAD/yake: https://github.com/fnl/segtok/blob/master/segtok/tokenizer.py#L301

**Comment by bunny-therapist** [comment]
Interestingly, syntok (https://github.com/fnl/syntok) is apparently the successor to segtok and it _does_ split "competition-centric" into two words like our current code does, unlike LIAAD/yake.

**Comment by xamgore** [comment]
I have started migrating `segtok`, completed segmenter for 80%, tokenizer is yet to do.

**Comment by bunny-therapist** [comment]
Nice!

**Comment by bunny-therapist** [comment]
Push the code where?

What exactly do you need? You want me to add tests?

**Comment by xamgore** [comment]
More like `assert!(python.tokenize(text) == rust.tokenize(text))`

**Comment by bunny-therapist** [comment]
Could just pre-tokenize with python and store that, like we did a bit higher up in this issue?

Of course, we could also just store texts and write tests which uses rust from python or python from rust.

**Comment by bunny-therapist** [comment]
Thanks. I starred it before there even was a commit XD
Would you like me to fo something with it or did the testing worl out?

**Comment by xamgore** [comment]
If I were you, I would concentrate on calling segtok from yake instead of unicode segmentation. Patch section in cargo.toml is needed. Tests might wait.

**Comment by bunny-therapist** [comment]
You mean I need to use patch to call segtok from yake because you have not published a crate for segtok?

**Comment by xamgore** [comment]
> Assuming that the list of tokens requested is the thing generated in [this line](https://github.com/LIAAD/yake/blob/master/yake/datarepresentation.py#L50):
> 
> ```
>  self.sentences_str = [ [w for w in split_contractions(web_tokenizer(s)) if not (w.startswith("'") and len(w) > 1) and len(w) > 0] for s in list(split_multi(text)) if len(s.strip()) > 0]     
> ```
> 
> Then you can find them in this attached file: [sentences_str.json](https://github.com/user-attachments/files/17586728/sentences_str.json)

I've tried it, the ported segtok has the same output as the original.

**Comment by xamgore** [comment]
> you have not published a crate

🎉🎉 https://crates.io/crates/segtok

**Comment by xamgore** [comment]
The lower [`sentences_str`](https://github.com/LIAAD/yake/blob/master/yake/datarepresentation.py#L50) can be built like this:

```rust
    let sentences: Vec<Vec<_>> = split_multi(input, Default::default())
        .into_iter()
        .filter(|span| !span.is_empty())
        .map(|span| {
            split_contractions(web_tokenizer(&span))
                .into_iter()
                .filter(|word| !(word.is_empty() || word.len() > 1 && word.starts_with("'")))
                .collect()
        })
        .collect();
```

The [upper](https://github.com/LIAAD/yake/blob/master/yake/datarepresentation.py#L33) is mostly the same.
